pkg/reconciler/apis/apibinding/apibinding_reconcile.go

```diff
@@ -237,209 +116,300 @@ func (r *workspaceAPIExportReferenceReconciler) reconcile(ctx context.Context, a
-			getCRD:               c.getCRD,
-			createCRD:            c.createCRD,
-			updateCRD:            c.updateCRD,
-			deletedCRDTracker:    c.deletedCRDTracker,
-		},
+func (c *controller) reconcileBound(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	apiExportClusterName, err := getAPIExportClusterName(apiBinding)
+	if err != nil {
+		// Should never happen
+		conditions.MarkFalse(
+			apiBinding,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportNotFoundReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			err.Error(),
+		)
+
+		return nil
 	}
 
-	var errs []error
+	if referencedAPIExportChanged(apiBinding) {
+		klog.Infof("APIBinding %s|%s needs rebinding because it now points to a different APIExport", apiBinding.ClusterName, apiBinding.Name)
 
-	for _, r := range reconcilers {
-		status, err := r.reconcile(ctx, apiBinding)
-		if err != nil {
-			errs = append(errs, err)
-		}
-		if status == reconcileStatusStop {
-			break
-		}
+		apiBinding.Status.Phase = apisv1alpha1.APIBindingPhaseBinding
+
+		return nil
 	}
 
-	return errors.NewAggregate(errs)
-}
+	apiExport, err := c.getAPIExport(apiExportClusterName, apiBinding.Spec.Reference.Workspace.ExportName)
+	if apierrors.IsNotFound(err) {
+		conditions.MarkFalse(
+			apiBinding,
+			apisv1alpha1.APIExportValid,
+			apisv1alpha1.APIExportNotFoundReason,
+			conditionsv1alpha1.ConditionSeverityWarning,
+			"APIExport %s|%s not found",
+			apiExportClusterName,
+			apiBinding.Spec.Reference.Workspace.ExportName,
+		)
 
-func (c *controller) getAPIExport(clusterName logicalcluster.LogicalCluster, name string) (*apisv1alpha1.APIExport, error) {
-	return c.apiExportsLister.Get(clusters.ToClusterAwareKey(clusterName, name))
-}
+		// Return nil here so we don't retry. If/when there is an informer event for the correct APIExport, this
+		// APIBinding will automatically be requeued.
+		return nil
+	}
+	if err != nil {
+		conditions.MarkFalse(
+			apiBinding,
+			apisv1alpha1.APIExportValid,
+			"TODO",
+			conditionsv1alpha1.ConditionSeverityWarning,
+			"Error getting APIExport %s|%s: %v",
+			apiExportClusterName,
+			apiBinding.Spec.Reference.Workspace.ExportName,
+			err,
+		)
 
-func (c *controller) getAPIResourceSchema(clusterName logicalcluster.LogicalCluster, name string) (*apisv1alpha1.APIResourceSchema, error) {
-	return c.apiResourceSchemaLister.Get(clusters.ToClusterAwareKey(clusterName, name))
-}
+		return err
+	}
 
-func (c *controller) getCRD(clusterName logicalcluster.LogicalCluster, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
-	return c.crdLister.Get(clusters.ToClusterAwareKey(clusterName, name))
-}
+	var exportedSchemas []*apisv1alpha1.APIResourceSchema
+	for _, schemaName := range apiExport.Spec.LatestResourceSchemas {
+		apiResourceSchema, err := c.getAPIResourceSchema(apiExportClusterName, schemaName)
+		if apierrors.IsNotFound(err) {
+			conditions.MarkFalse(
+				apiBinding,
+				apisv1alpha1.APIExportValid,
+				"TODO",
```

Comment:
APIExportInvalidSchemas

Reply:
Do we want to expose this information to the user?

Reply:
not necessary I think